File: apps/api/app/gemini_live.py
from __future__ import annotations

import asyncio
import inspect
import logging
import traceback

from google import genai
from google.genai import types

logger = logging.getLogger(__name__)

ACTIVITY_START = object()
ACTIVITY_END = object()
AUDIO_STREAM_END = object()


class GeminiLive:
    def __init__(
        self,
        api_key: str,
        model: str,
        system_instruction: str,
        tools: list[types.Tool],
        tool_mapping: dict,
        input_sample_rate: int = 16000,
    ) -> None:
        self.api_key = api_key
        self.model = model
        self.system_instruction = system_instruction
        self.tools = tools
        self.tool_mapping = tool_mapping
        self.input_sample_rate = input_sample_rate
        self.client = genai.Client(api_key=api_key)

    async def start_session(
        self,
        audio_input_queue: asyncio.Queue,
        audio_output_callback,
        audio_interrupt_callback=None,
        order_update_callback=None,
    ):
        config = types.LiveConnectConfig(
            response_modalities=[types.Modality.AUDIO],
            speech_config=types.SpeechConfig(
                voice_config=types.VoiceConfig(
                    prebuilt_voice_config=types.PrebuiltVoiceConfig(voice_name="Aoede")
                )
            ),
            system_instruction=types.Content(
                parts=[types.Part(text=self.system_instruction)]
            ),
            input_audio_transcription=types.AudioTranscriptionConfig(),
            output_audio_transcription=types.AudioTranscriptionConfig(),
            tools=self.tools,
        )

        logger.info("Connecting to Gemini Live model=%s", self.model)
        async with self.client.aio.live.connect(model=self.model, config=config) as session:
            logger.info("Gemini Live session opened")

            async def send_audio() -> None:
                try:
                    while True:
                        chunk = await audio_input_queue.get()
                        if chunk is ACTIVITY_START:
                            await session.send_realtime_input(
                                activity_start=types.ActivityStart()
                            )
                            logger.debug("Sent activity_start")
                        elif chunk is ACTIVITY_END:
                            await session.send_realtime_input(
                                activity_end=types.ActivityEnd()
                            )
                            logger.debug("Sent activity_end")
                        elif chunk is AUDIO_STREAM_END:
                            await session.send_realtime_input(audio_stream_end=True)
                            logger.info("Sent audio_stream_end")
                        elif isinstance(chunk, (bytes, bytearray)) and len(chunk) > 0:
                            await session.send_realtime_input(
                                audio=types.Blob(
                                    data=bytes(chunk),
                                    mime_type=f"audio/pcm;rate={self.input_sample_rate}",
                                )
                            )
                        else:
                            logger.debug("Skipped empty audio chunk")
                except asyncio.CancelledError:
                    logger.debug("send_audio cancelled")
                except Exception as exc:  # noqa: BLE001
                    logger.error("send_audio error: %s\n%s", exc, traceback.format_exc())

            event_queue: asyncio.Queue = asyncio.Queue()

            async def receive_loop() -> None:
                try:
                    while True:
                        async for response in session.receive():
                            server_content = response.server_content
                            tool_call = response.tool_call

                            if server_content:
                                if server_content.model_turn:
                                    for part in server_content.model_turn.parts:
                                        if part.inline_data and part.inline_data.data:
                                            if inspect.iscoroutinefunction(audio_output_callback):
                                                await audio_output_callback(part.inline_data.data)
                                            else:
                                                audio_output_callback(part.inline_data.data)

                                if (
                                    server_content.input_transcription
                                    and server_content.input_transcription.text
                                ):
                                    text = server_content.input_transcription.text.rstrip()
                                    if text.strip():
                                        logger.info("User transcript: %s", text)
                                        await event_queue.put(
                                            {
                                                "type": "transcript.user",
                                                "text": text,
                                            }
                                        )

                                if (
                                    server_content.output_transcription
                                    and server_content.output_transcription.text
                                ):
                                    text = server_content.output_transcription.text.rstrip()
                                    if text.strip():
                                        logger.info("Assistant transcript: %s", text)
                                        await event_queue.put(
                                            {
                                                "type": "transcript.assistant",
                                                "text": text,
                                            }
                                        )

                                if server_content.turn_complete:
                                    await event_queue.put({"type": "turn_complete"})

                                if server_content.interrupted:
                                    if audio_interrupt_callback:
                                        if inspect.iscoroutinefunction(audio_interrupt_callback):
                                            await audio_interrupt_callback()
                                        else:
                                            audio_interrupt_callback()
                                    await event_queue.put({"type": "interrupted"})

                            if tool_call:
                                function_responses = []
                                for fc in tool_call.function_calls:
                                    func_name = fc.name
                                    args = dict(fc.args or {})
                                    result: dict | str

                                    if func_name in self.tool_mapping:
                                        try:
                                            tool_func = self.tool_mapping[func_name]
                                            if inspect.iscoroutinefunction(tool_func):
                                                result = await tool_func(**args)
                                            else:
                                                loop = asyncio.get_running_loop()
                                                result = await loop.run_in_executor(
                                                    None, lambda: tool_func(**args)
                                                )
                                        except Exception as exc:  # noqa: BLE001
                                            result = {"error": str(exc)}
                                    else:
                                        result = {"error": f"Unknown tool '{func_name}'."}

                                    function_responses.append(
                                        types.FunctionResponse(
                                            name=func_name,
                                            id=fc.id,
                                            response={"result": result},
                                        )
                                    )

                                    await event_queue.put(
                                        {
                                            "type": "tool_call",
                                            "name": func_name,
                                            "args": args,
                                            "result": result,
                                        }
                                    )

                                    if order_update_callback and isinstance(result, dict):
                                        order = result.get("order")
                                        if order:
                                            await order_update_callback(order)

                                await session.send_tool_response(
                                    function_responses=function_responses
                                )
                except asyncio.CancelledError:
                    logger.debug("receive_loop cancelled")
                except Exception as exc:  # noqa: BLE001
                    logger.error("receive_loop error: %s\n%s", exc, traceback.format_exc())
                    await event_queue.put({"type": "error", "error": str(exc)})
                finally:
                    await event_queue.put(None)

            send_audio_task = asyncio.create_task(send_audio())
            receive_task = asyncio.create_task(receive_loop())

            try:
                while True:
                    event = await event_queue.get()
                    if event is None:
                        break
                    yield event
            finally:
                send_audio_task.cancel()
                receive_task.cancel()

File: apps/server/src/services/gemini-live.ts
import { GoogleGenAI, Modality } from "@google/genai";
import { formatConnectionError, withDirectConnectionAsync } from "./networking.js";
import { buildToolDeclarations, buildToolMapping, type ProductInfo } from "./tools.js";
import { buildBookingToolMapping } from "./booking-tools.js";
import { buildFaqToolMapping } from "./faq-tools.js";
import type { OrderStore } from "./order-store.js";

export const ACTIVITY_START = Symbol("ACTIVITY_START");
export const ACTIVITY_END = Symbol("ACTIVITY_END");
export const AUDIO_STREAM_END = Symbol("AUDIO_STREAM_END");
export const SESSION_RECONNECT = Symbol("SESSION_RECONNECT");
export const SHUTDOWN = Symbol("SHUTDOWN");

export class ClientTextEvent {
  constructor(public text: string) {}
}

const MAX_SESSION_RECONNECTS = 8;
const RECOVERABLE_MARKERS = [
  "keepalive ping timeout",
  "GoAway",
  "go away",
  "service is currently unavailable",
  "ConnectionClosed",
];

type SessionEvent = Record<string, unknown> | typeof SESSION_RECONNECT | null;

function isRecoverable(exc: unknown): boolean {
  const message = String(exc);
  return RECOVERABLE_MARKERS.some((marker) => message.toLowerCase().includes(marker.toLowerCase()));
}

export interface GeminiLiveOptions {
  apiKey: string;
  model: string;
  systemInstruction: string;
  orderStore: OrderStore;
  products: ProductInfo[];
  orderingEnabled?: boolean;
  bookingEnabled?: boolean;
  faqEnabled?: boolean;
  businessId?: string;
  voiceSessionId?: string;
  onConfirm?: (order: Record<string, unknown>) => void;
  onSetCustomerName?: (name: string) => void;
  inputSampleRate?: number;
}

export async function* startGeminiSession(
  options: GeminiLiveOptions,
  audioInputQueue: AsyncIterable<unknown>,
  callbacks: {
    audioOutput: (data: Buffer) => Promise<void>;
    audioInterrupt?: () => Promise<void>;
    orderUpdate?: (order: Record<string, unknown>) => Promise<void>;
  },
): AsyncGenerator<Record<string, unknown>> {
  const toolMapping = options.bookingEnabled
    ? buildBookingToolMapping({
        businessId: options.businessId ?? "",
        products: options.products,
        voiceSessionId: options.voiceSessionId,
      })
    : options.faqEnabled
      ? buildFaqToolMapping({})
      : buildToolMapping(options.orderStore, options.products, {
          onConfirm: options.onConfirm,
          onSetCustomerName: options.onSetCustomerName,
          orderingEnabled: options.orderingEnabled ?? true,
        });

  let reconnects = 0;
  while (reconnects <= MAX_SESSION_RECONNECTS) {
    if (reconnects > 0) {
      yield { type: "session.status", status: "reconnecting" };
      await new Promise((r) => setTimeout(r, 400));
    }

    try {
      yield* runSingleSession(options, audioInputQueue, callbacks, toolMapping);
      return;
    } catch (exc) {
      if (isRecoverable(exc)) {
        reconnects += 1;
        if (reconnects > MAX_SESSION_RECONNECTS) {
          yield {
            type: "error",
            error: "Sesi suara terputus. Ketuk Order Now untuk menyambung lagi.",
          };
          return;
        }
        continue;
      }
      throw exc;
    }
  }
}

type ToolHandlerResult = Record<string, unknown> | Promise<Record<string, unknown>>;
type ToolMapping = Record<string, (args: Record<string, unknown>) => ToolHandlerResult>;

async function* runSingleSession(
  options: GeminiLiveOptions,
  audioInputQueue: AsyncIterable<unknown>,
  callbacks: {
    audioOutput: (data: Buffer) => Promise<void>;
    audioInterrupt?: () => Promise<void>;
    orderUpdate?: (order: Record<string, unknown>) => Promise<void>;
  },
  toolMapping: ToolMapping,
): AsyncGenerator<Record<string, unknown>> {
  const sampleRate = options.inputSampleRate ?? 16000;
  const eventQueue: SessionEvent[] = [];
  let resolveEvent: (() => void) | null = null;
  let sessionClosed = false;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let liveSession: any = null;

  const pushEvent = (event: SessionEvent) => {
    eventQueue.push(event);
    resolveEvent?.();
    resolveEvent = null;
  };

  const waitEvent = (): Promise<void> =>
    new Promise((resolve) => {
      if (eventQueue.length) {
        resolve();
        return;
      }
      resolveEvent = resolve;
    });

  const sendAudioLoop = async () => {
    try {
      for await (const chunk of audioInputQueue) {
        if (!liveSession || sessionClosed) break;
        if (chunk === SHUTDOWN) {
          pushEvent(null);
          break;
        }
        if (chunk === ACTIVITY_START) {
          await liveSession.sendRealtimeInput({ activityStart: {} });
        } else if (chunk === ACTIVITY_END) {
          await liveSession.sendRealtimeInput({ activityEnd: {} });
        } else if (chunk === AUDIO_STREAM_END) {
          await liveSession.sendRealtimeInput({ audioStreamEnd: true });
        } else if (chunk instanceof ClientTextEvent) {
          await liveSession.sendClientContent({
            turns: [{ role: "user", parts: [{ text: chunk.text }] }],
            turnComplete: true,
          });
        } else if (chunk instanceof Buffer || chunk instanceof Uint8Array) {
          if (chunk.length > 0) {
            await liveSession.sendRealtimeInput({
              audio: {
                data: Buffer.from(chunk).toString("base64"),
                mimeType: `audio/pcm;rate=${sampleRate}`,
              },
            });
          }
        }
      }
    } catch {
      // cancelled
    }
  };

  await withDirectConnectionAsync(async () => {
    const ai = new GoogleGenAI({ apiKey: options.apiKey });
    liveSession = await ai.live.connect({
      model: options.model,
      config: {
        responseModalities: [Modality.AUDIO],
        speechConfig: {
          voiceConfig: { prebuiltVoiceConfig: { voiceName: "Aoede" } },
        },
        systemInstruction: options.systemInstruction,
        inputAudioTranscription: {},
        outputAudioTranscription: {},
        tools: buildToolDeclarations({
          orderingEnabled: options.orderingEnabled ?? true,
          bookingEnabled: options.bookingEnabled ?? false,
          faqEnabled: options.faqEnabled ?? false,
        }) as never,
      },
      callbacks: {
        onmessage: async (message) => {
          if (sessionClosed) return;

          if (message.goAway) {
            pushEvent(SESSION_RECONNECT);
            return;
          }

          const serverContent = message.serverContent;
          const toolCall = message.toolCall;

          if (serverContent) {
            if (serverContent.modelTurn?.parts) {
              for (const part of serverContent.modelTurn.parts) {
                if (part.inlineData?.data) {
                  const data = Buffer.from(part.inlineData.data, "base64");
                  await callbacks.audioOutput(data);
                }
              }
            }

            if (serverContent.inputTranscription?.text?.trim()) {
              pushEvent({
                type: "transcript.user",
                text: serverContent.inputTranscription.text.trimEnd(),
              });
            }

            if (serverContent.outputTranscription?.text?.trim()) {
              pushEvent({
                type: "transcript.assistant",
                text: serverContent.outputTranscription.text.trimEnd(),
              });
            }

            if (serverContent.turnComplete) {
              pushEvent({ type: "turn_complete" });
            }

            if (serverContent.interrupted) {
              await callbacks.audioInterrupt?.();
              pushEvent({ type: "interrupted" });
            }
          }

          if (toolCall?.functionCalls) {
            const functionResponses = [];
            for (const fc of toolCall.functionCalls) {
              const funcName = fc.name ?? "";
              const args = (fc.args ?? {}) as Record<string, unknown>;
              let result: Record<string, unknown> | string;

              if (toolMapping[funcName]) {
                try {
                  const rawResult = toolMapping[funcName](args);
                  result =
                    rawResult instanceof Promise ? await rawResult : (rawResult as Record<string, unknown>);
                } catch (exc) {
                  result = { error: String(exc) };
                }
              } else {
                result = { error: `Unknown tool '${funcName}'.` };
              }

              functionResponses.push({
                name: funcName,
                id: fc.id,
                response: { result },
              });

              pushEvent({
                type: "tool_call",
                name: funcName,
                args,
                result,
              });

              if (callbacks.orderUpdate && typeof result === "object" && result && "order" in result) {
                await callbacks.orderUpdate(result.order as Record<string, unknown>);
              }
            }

            await liveSession?.sendToolResponse({ functionResponses });
          }
        },
        onerror: (e) => {
          if (isRecoverable(e)) {
            pushEvent(SESSION_RECONNECT);
          } else {
            pushEvent({ type: "error", error: formatConnectionError(e) });
            pushEvent(null);
          }
        },
        onclose: () => {
          pushEvent(null);
        },
      },
    });
  });

  const sendTask = sendAudioLoop();

  try {
    while (true) {
      await waitEvent();
      const event = eventQueue.shift();
      if (event === SESSION_RECONNECT) {
        throw new Error("Gemini session ended");
      }
      if (event === null) break;
      if (event && typeof event === "object") {
        yield event as Record<string, unknown>;
      }
    }
  } finally {
    sessionClosed = true;
    try {
      liveSession?.close();
    } catch {
      // ignore
    }
    await sendTask.catch(() => undefined);
  }
}

export function createAudioQueue(): {
  push: (item: unknown) => void;
  iterable: AsyncIterable<unknown>;
} {
  const queue: unknown[] = [];
  let resolveNext: ((value: IteratorResult<unknown>) => void) | null = null;
  let done = false;

  const iterable: AsyncIterable<unknown> = {
    [Symbol.asyncIterator]() {
      return {
        async next(): Promise<IteratorResult<unknown>> {
          if (queue.length) {
            return { value: queue.shift()!, done: false };
          }
          if (done) return { value: undefined, done: true };
          return new Promise((resolve) => {
            resolveNext = resolve;
          });
        },
      };
    },
  };

  return {
    push(item: unknown) {
      if (resolveNext) {
        resolveNext({ value: item, done: false });
        resolveNext = null;
      } else {
        queue.push(item);
      }
      if (item === SHUTDOWN) done = true;
    },
    iterable,
  };
}
